### PR TITLE
Test refactor

### PR DESCRIPTION
**Pull request by GBuella**
Introduce mock_tx_alloc, and decrease LARGE_LIMIT.
#50

### CODE REVIEW

**Comment by GBuella** [comment]
It might be useful if @pbalcer took a look at the allocation mocking. As I understand, pmemobj_tx_alloc is his child.

File: src/mock_tx_alloc.cc [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    static PMEMoid (*actual)(size_t, uint64_t);
+
+    if (actual == nullptr) {
+        actual = (decltype(actual))dlsym(RTLD_NEXT, "pmemobj_tx_alloc");
+        if (actual == nullptr)
+            abort();
+    }
+
+    if (tx_alloc_should_fail)
+        return OID_NULL;
```
Comment (pbalcer) [inline src/mock_tx_alloc.cc]: set the errno to ENOMEM.

File: src/mock_tx_alloc.cc [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#include <cstdlib>
+
+#include "mock_tx_alloc.h"
+
+thread_local bool tx_alloc_should_fail;
+
+extern "C" PMEMoid pmemobj_tx_alloc(size_t size, uint64_t type_num);
+
+PMEMoid pmemobj_tx_alloc(size_t size, uint64_t type_num) {
+    static PMEMoid (*actual)(size_t, uint64_t);
```
Comment (pbalcer) [inline src/mock_tx_alloc.cc]: typedef PMEMoid real(size_t, uint64_t);

real *real_func = (real *)dlsym(RTLD_NEXT, ...);

isn't that easier and a bit cleaner (you don't omit the pointer types near the function).

Comment (tomaszkapela) [inline src/mock_tx_alloc.cc]: :+1: on the typedef, however my preference would be to add the pointer to the typedef. Up to you though.

File: src/mock_tx_alloc.cc [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    if (actual == nullptr) {
```
Comment (tomaszkapela) [inline src/mock_tx_alloc.cc]: This is not thread safe and it probably should be.

File: src/mock_tx_alloc.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
```
Comment (tomaszkapela) [inline src/mock_tx_alloc.h]: `#pragma`? Rly?

File: src/mock_tx_alloc.cc (hunk 3 above)
Comment (GBuella) [inline src/mock_tx_alloc.cc]: It happens to be, as a pointer is read / written atomically on x86_64 anyways.
But I already started to work on making it more clear:
https://github.com/GBuella/pmemkv/commit/32ff7531ca76372a030cbd53a85643c057e26630

File: src/mock_tx_alloc.h (hunk 4 above)
Comment (GBuella) [inline src/mock_tx_alloc.h]: That is what I see in `pmemkv.h`, so I just followed that convention.

File: src/mock_tx_alloc.cc (hunk 3 above)
Comment (GBuella) [inline src/mock_tx_alloc.cc]: Done

File: src/mock_tx_alloc.cc (hunk 1 above)
Comment (GBuella) [inline src/mock_tx_alloc.cc]: Done

File: src/mock_tx_alloc.cc (hunk 2 above)
Comment (GBuella) [inline src/mock_tx_alloc.cc]: I guess no typedef is needed, `decltype(pmemobj_tx_alloc)` should be enough.

File: src/mock_tx_alloc.h (hunk 4 above)
Comment (RobDickinson) [inline src/mock_tx_alloc.h]: ...and I think I copied `#pragma` from RocksDB originally (the very first pmemkv API was a RocksDB clone).  :grin: 

I don't have strong feelings about using this or not, but @GBuella sets a good example in keeping things consistent.  If we're gonna remove the `#pragma once` for good reason then let's consider all the cases.

Comment (tomaszkapela) [inline src/mock_tx_alloc.h]: Pretty much every compiler supports this. It's just that you rarely see this on Linux, that's why I asked. Consistency is a good reason to keep it that way.

**Pull request merged**